pkg/providers/docker/controlplane_test.go

```diff
@@ -0,0 +1,566 @@
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/providers/docker"
+	"github.com/aws/eks-anywhere/pkg/utils/ptr"
+	releasev1alpha1 "github.com/aws/eks-anywhere/release/api/v1alpha1"
+)
+
+func TestControlPlaneObjects(t *testing.T) {
+	tests := []struct {
+		name         string
+		controlPlane *docker.ControlPlane
+		want         []kubernetes.Object
+	}{
+		{
+			name: "stacked etcd",
+			controlPlane: &docker.ControlPlane{
+				Cluster:                     capiCluster(),
+				ProviderCluster:             dockerCluster(),
+				KubeadmControlPlane:         kubeadmControlPlane(),
+				ControlPlaneMachineTemplate: dockerMachineTemplate("cp-mt"),
+			},
+			want: []kubernetes.Object{
+				capiCluster(),
+				dockerCluster(),
+				kubeadmControlPlane(),
+				dockerMachineTemplate("cp-mt"),
+			},
+		},
+		{
+			name: "unstacked etcd",
+			controlPlane: &docker.ControlPlane{
+				Cluster:                     capiCluster(),
+				ProviderCluster:             dockerCluster(),
+				KubeadmControlPlane:         kubeadmControlPlane(),
+				ControlPlaneMachineTemplate: dockerMachineTemplate("cp-mt"),
+				EtcdCluster:                 etcdCluster(),
+				EtcdMachineTemplate:         dockerMachineTemplate("etcd-mt"),
+			},
+			want: []kubernetes.Object{
+				capiCluster(),
+				dockerCluster(),
+				kubeadmControlPlane(),
+				dockerMachineTemplate("cp-mt"),
+				etcdCluster(),
+				dockerMachineTemplate("etcd-mt"),
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			g.Expect(tt.controlPlane.Objects()).To(ConsistOf(tt.want))
+		})
+	}
+}
+
+func TestControlPlaneSpecNewCluster(t *testing.T) {
+	g := NewWithT(t)
+	logger := test.NewNullLogger()
+	ctx := context.Background()
+	client := test.NewFakeKubeClient()
+	spec := testClusterSpec()
+
+	cp, err := docker.ControlPlaneSpec(ctx, logger, client, spec)
+
+	wantCPMachineTemplate := dockerMachineTemplate("test-control-plane-1")
+	wantEtcdMachineTemplate := dockerMachineTemplate("test-etcd-1")
+
+	g.Expect(err).NotTo(HaveOccurred())
+	g.Expect(cp).NotTo(BeNil())
+	g.Expect(cp.Cluster).To(Equal(capiCluster()))
+	g.Expect(cp.KubeadmControlPlane).To(Equal(kubeadmControlPlane()))
+	g.Expect(cp.EtcdCluster).To(Equal(etcdCluster()))
+	g.Expect(cp.ProviderCluster).To(Equal(dockerCluster()))
+	g.Expect(cp.ControlPlaneMachineTemplate).To(Equal(wantCPMachineTemplate))
+	g.Expect(cp.EtcdMachineTemplate).To(Equal(wantEtcdMachineTemplate))
+}
+
+func TestControlPlaneSpecNoKubeVersion(t *testing.T) {
+	g := NewWithT(t)
+	logger := test.NewNullLogger()
+	ctx := context.Background()
+	client := test.NewFakeKubeClient()
+	spec := testClusterSpec()
+	spec.Cluster.Spec.KubernetesVersion = ""
+
+	_, err := docker.ControlPlaneSpec(ctx, logger, client, spec)
+	g.Expect(err).To(MatchError(ContainSubstring("generating docker control plane yaml spec")))
+}
+
+func TestControlPlaneSpecUpdateMachineTemplates(t *testing.T) {
```

Comment:
Could we have a test for the scenario of an upgrade without changes in the machine templates?
We recently added one of those to sphere if you want to check it out :)

I think it's important, we missed it for vsphere and that's why we didn't the bug earlier.

Reply:
Added this. `DockerMachineTemplate` does not have a Snapshot field, but I used the `ProviderID` field instead to test defaults.